constraint/pkg/client/client_test.go

```diff
@@ -465,13 +467,85 @@ func TestRemoveTemplate(t *testing.T) {
 			if err != nil {
 				t.Fatal(err)
 			}
+			_, err = c.AddTemplate(context.Background(), tt.Template)
+			if err != nil && !tt.ErrorExpected {
+				t.Errorf("err = %v; want nil", err)
+			}
+			if err == nil && tt.ErrorExpected {
+				t.Error("err = nil; want non-nil")
+			}
 			r, err := c.RemoveTemplate(context.Background(), tt.Template)
+			if err != nil {
+				t.Errorf("err = %v; want nil", err)
+			}
+			expectedCount := 0
+			expectedHandled := make(map[string]bool)
+			if !tt.ErrorExpected {
+				expectedCount = 1
+				expectedHandled = map[string]bool{"h1": true}
+			}
+			if r.HandledCount() != expectedCount {
```

Comment:
It's not clear whether `r` is valid and non-nil on this line if we received an error from `RemoveTemplate` above.

Reply:
`r` is always returned as other handlers may have had the chance to process before an error was raised.

Reply:
If that's the case I might suggest some combination of these:

* Returning the responses as a value rather than a pointer
* Checking for nil receiver in `HandledCount` and `Results`
* Documenting that responses is valid even upon error

I generally assume return values are invalid when an error is returned, which is also echoed here: https://github.com/golang/go/wiki/CodeReviewComments#in-band-errors

>  Instead of requiring clients to check for an in-band error value, a function should return an additional value to indicate whether its other return values are valid. This return value may be an error, or a boolean when no explanation is needed. It should be the final return value.

...but I recognize the need to report additional information on partial success.

Reply:
Re: value instead of pointer...

Given that the amount of memory it takes to represent this object potentially scales with the number of responses, I'm wary of using a value. Since it's currently represented by a map, that may not be an issue, though I like the idea of guaranteeing we are only ever copying-by-reference.

Re: checking for a nil receiver...

That's a good idea, then the helper functions could return empty results

Re: documentation... ya, there is definitely interface documentation work to be done

Looking through that doc, the "how to do errors" gets a bit hand-wavy at points b/c they can't predict all uses. I think this is one. I think what we have strikes a reasonable balance between the two competing needs and follows the main point, which is "report errors out-of-band to avoid the need for introspection"

Note that if a user is not interested in the result if an error is returned, this devolves into standard Go code.

Reply:
Sounds reasonable 👍🏻

Reply:
Added the checking for a nil receiver and documented that `results` are still valid even when error is returned.

Reply:
good call out!